Created by Ariella Gladstein, based on code from Consuelo Quinto Cortes and Krishna Veeramah.
Also worked on by Mack Skodiak, David Christy, and Logan Gantner.  
agladstein@email.arizona.edu

## About
The main script `simprily.py` runs genome simulations with user defined parameters or parameters randomly generated by priors and computes genomic statistics on the simulation output.  
Version 1

1. Run genome simulation with model defined by prior distributions of parameters and demographic model structure.
2. Take into account SNP array ascrtainment bias by creating pseudo array based on priors of number of samples of discovery populations and allele frequency cut off.
3. Calculate genomic summary statistics on simulated genomes and pseudo arrays. 

This is ideal for use with Approximate Bayesian Computation on whole genome or SNP array data.

Uses c++ programs macs and GERMLINE. For more information on these programs, see:  
https://github.com/gchen98/macs  
https://github.com/sgusev/GERMLINE  

## Install
cd to the directory you want to work in,
```bash
git clone https://github.com/agladstein/SimPrily.git
```


## Environment Set up
Python 2.7.6 or 2.7.11 is required to run the code, with the requirements installed from requirements.txt. The most reliable way to do this is with a virtual environment. *In the future I will likely include a docker image*.


If using Vagrant:

```bash
vagrant up
vagrant ssh
```

```bash
sudo apt-get update
sudo apt-get upgrade
sudo apt-get install python-virtualenv git python-dev
sudo easy_install -U distribute
cd ~
virtualenv simprily_env
source ~/simprily_env/bin/activate
pip install --upgrade pip
pip install pip-tools
cd /vagrant
pip-sync
~/simprily_env/bin/python simprily.py macs 1 array_template/ill_650_test.bed 1000000 123456 prior 0
```

If not using Vagrant:
```bash
sudo apt-get update
sudo apt-get upgrade
sudo apt-get install python-virtualenv git python-dev
sudo easy_install -U distribute
virtualenv simprily_env
pip install --upgrade pip
pip install pip-tools
pip-sync
simprily_env/bin/python simprily.py examples/eg2/param_file_eg2.txt examples/eg2/model_file_eg2.csv macs 1 array_template/ill_650_test.bed 0 True output_dir
```

If the above options do not work, the correct version of Python can also be installed locally:  

```bash
cd mkdir python_prebuild
wget https://www.python.org/ftp/python/2.7.6/Python-2.7.6.tgz
mkdir python
tar -zxvf Python-2.7.6.tgz
cd Python-2.7.6
./configure --prefix=$(pwd)/../python
make
make install
cd ..
export PATH=$(pwd)/python/bin:$PATH
wget https://bootstrap.pypa.io/get-pip.py
python get-pip.py
pip install -r /home/agladstein/simslg/macsswig_simslg/requirements.txt
python simprily.py examples/eg2/param_file_eg2.txt examples/eg2/model_file_eg2.csv macsSwig 1 array_template/ill_650_test.bed 0 True output_dir
```

## Usage
e.g. One Full simulation:  
```
python simprily.py examples/eg3/param_file_eg3.txt examples/eg3/model_file_eg3.csv macsSwig 1 array_template/Axiom_LAT_chr1.bed 0 True output_dir
```

e.g. One Test simulation:  
```
python simprily.py examples/eg3/param_file_eg3.txt examples/eg3/model_file_eg3.csv macs 1 array_template/ill_650_test.bed 0 True out_edit2
```

#### Input  
`simprily.py` takes 8 arguments.   

Run as  
```
python simprily.py param_file.txt model_file.csv sim_option jobID array_template germline output_dir
```
1. `param_file.txt` = full path to file containing parameter values or priors
2. `model_file.csv` = full path to file containing model commands
3. `sim_option` = macs or macsswig
4. `jobID` = can be any unique value to identify the output  
5. `array_template` = bed file with physical position of SNPs on array to use as template for simulated pseudo array.  
6. `germline` = 0 to run GERMLINE, 1 to not run GERMLINE.  
7. `random_discovery` = True to randomly pick number of individuals for SNP discovery, or False to use half of the discovery individuals.
8. `output_dir` = path to the directory to output to. No argument will use the default of current dir `.`

### Additional information on input arguments

#### Required arguments

##### param_file.txt
Examples of param_file.txt can be found in examples.
The param_file.txt must define the parameters of the demographic model and the minimum derived allele frequency to be used to create the pseudo array, if a pseudo array is to be created.

All parameter values should be given in pre-coalescent scaled units. 
That is, Ne should be given in units of chromosomes, and time should be given in units of generations.
The code will scale to the appropriate coalescent units for the simulation.

The definition can be hard-coded parameter values, such as:
```text
A = 1000
B = 1000
T1 = 100
```

The definition can be a prior, such as:  
```text
A = (1e3.0:1e4.0)
B = (1e3.0:1e4.0)
T1 = (10:500)
```
Log base 10 can be used for the parameter definitions by using `1eX` or `1Ex`.
This is recommended when using a prior with a very large range (See ABCtoolbox manual).

If pseudo arrays are to be created, the derived allele frequency must be defined. For example,
```text
A = (1e3.0:1e4.0)
B = (1e3.0:1e4.0)
T1 = (10:500)
daf = (0.01:0.1)
```

*currently only a range of values is supported for daf. Therefore if you want to hard code a value, use the same value as the min and max of the prior.*

##### model_file.csv
Examples of param_file.txt can be found in examples.

The demographic and SNP ascertainment models are defined in the model_file.csv.

All instances of any argument must start with a dash followed by the corresponding argument parameters, 
and value(s). 
Each new argument must be a new line.
All variables and values must be separated by commas (white space will be ignored, so it is okay to include spaces).
The model arguments can appear in any order.

All parameters must be called with a name corresponding to its definition in the param file.
This is how parameter values are assigned to the simulation model.
For example,
```text
-macs,./bin/macs,
-length,5000000,
-s,1231414,
-t,2.5e-8,
-r,1e-8,
-h,1e5,
-R,genetic_map_b37/genetic_map_GRCh37_chr1.txt.macshs,
-I, 2, 50, 50
-n, 1, A
-n, 2, B
-ej, T1, 1, 2
```

###### Setup simulation arguments
`-macs`: macs location  
* Path to macs executable in relation to the working directory.
This is only relavent if `macs` is specified in the command line argument, since `macs_file` does not simulate and `macsswig` uses it's own internal macs version. 
When using the `macs` option, we recommend using the compiled executable in `./macs`.

`-length`: length  
* how many base pairs you want to simulate

`-s`  
* random seed. 
Must be an integer.  
If no input is given, no seed will be used, and everything will be random.
If a seed is provided, reproducible parameters will be picked from the priors.
Using a seed will also cause reproducible simulations with macs.  
*The seed does not currently work with macsswig*

###### Demographic simulation arguments
All argument flags are based on macs arguments (see macs manual for more detail).  

`-t`  
* mutation rate per site per 4N generations



`-d`  
* enable debugging messages. No entry will default to allowing debugging 
messages. This will not work when using macsswig

`-h`  
* history. Refers to the number of previous base pairs to retain

`-r`: r    
* recombination rate per site per 4N generations

`-c`: f lambda  
* f = ratio of gene conversion rate to crossover rate. track len(lambda) 
is mean length of tract in base pairs.
*This has not been tested.*

`-R`  
* Tab delimited file where first two columns indicate range of base 
pair positions scaled to the unit interval and last column is ratio 
with respect to base line recombination rate.
We provide the HapMap genetic map for each chromosome in genetic_map_b37.

`-T`  
* Print each local tree in Newick format to standard out. *This has not been tested.*

`-G`: alpha 
* Assign growth rate alpha across populations where 
alpha=-log(Np/Nr)
* alpha will accept priors in the format a;b where a is the 
lower value and b is the higher value

`-I`: n n_n  
* Assign all elements of the migration matrix for n populations. 
Values in matrix set to mig_rate/(n-1)
* The length of n_n should be equal to n

`-m`: i,j m
* i, j is associated with a location in the migration matrix
* m is assigned to the value at (i, j)

`-ma`: m_nn
* Assign values to all elements of migration matrix for n populations

`-n`: i size
* Population i set to size

`-g`: i alpha 
* assigns alpha value as explained in -G to population i

`-eG`: t alpha
* t is a time value.
* alpha behaves the same as in -G

`-eg`: t i alpha
* t is a time value. 
* alpha behaves the same as in -G
* i is a population that alpha is assigned to at time t

`-eM`: t m 
* t is a time value. 
* Assign migration rate m to all elements in migration matrix at 
time t 

`-em`: t i,j m_ij
* t is a time value. 
* i and j make up point in a population matrix. 
* assigns migration rate m_ij to the population at i, j at time t 

`-ema`: t n m_nn
* t is a time value. 
* Assign migration rates  within the migration matrix for n 
populations at time t

`-eN`: t size
* t is a time value. 
* Assigns size to all populations at time t

`-en`: t i size_i
* t is a time value. 
* assigns size_i to population i at time t

`-es`: t i p
* t is a time value. 
* splits population i by p at time t

`-ej`: t i j
* t is a time value. 
* joins population i with population j at time t

###### SNP array ascertainment arguments
If the user would like to create a pseudo array from the simulation, three additional arguments must be included in the model_file:  
`-discovery`, followed by the populations (defined by their numbers from `-n`) that should be used to discover the SNP (e.g. the HapMap populations).
These are the populations that will be used to create the pseudo array.
When calculating summary statistics, summary statistics based on whole genome simulation and pseudo array will be calculated for these populations.    
`-sample`, followed by the populations (defined by their numbers from `-n`) that are the samples of interest for demographic interest.      
`daf`, followed by the parameter name for daf.

For example:
```text
-macs,./bin/macs,
-length,5000000,
-s,1231414,
-t,2.5e-8,
-r,1e-8,
-h,1e5,
-R,genetic_map_b37/genetic_map_GRCh37_chr1.txt.macshs,
-I, 2, 50, 50
-n, 1, A
-n, 2, B
-ej, T1, 1, 2
-discovery, 1
-sample, 2
-daf, daf
```

###### Ordering of time-specific events
When using priors, if some demographic events must happen in a certain order, the order can be specified by adding the order number to the argument.
For example say there are two demographic events, a population split and instantaneous growth, but the instantaneous growth must happen before the population split, we can indicate that in the model file:
```text
-en_1, Tgrowth, 1, A2
-ej_2, Tsplit, 2, 1 
```

Additionally, the same format can be used to indicate that multiple events should happen at the same time.
If there are multiple events that should happen at the same time, the word `inst` should be used instead of a time parameter after the first definition of the time.
*(this will actually cause the times to be just different enough that macs is happy.)*  
For example, say we wanted growth to occur at the same time as the population split:
```text
-en_1, Tgrowth, 1, A2
-ej_1, inst, 2, 1 
```
In this case, the population split will technically be simulated slightly after the growth.

##### sim_option
Choose between `macs` and `macsswig`.  
`macs_file` - read in static output from MaCS. This should only be used for rigorous testing.  
`macs` - use the original simulator [MaCS](https://github.com/gchen98/macs). This option will stream the MaCS simulation output directly to be read into a python bitarray. This option is ideal for rigorous testing that requires stable reproducible results.  
`macsswig` - use the modified version of MaCS, macsswig.

##### jobID
This is a unique identifier for the job. It is used in the names of the output files.
For example, the output file with the summary statistics is named `ms_output_{jobid}.summary`.

##### random_discovery
`True` to randomly pick number of individuals for SNP discovery, or `False` to use all discovery individuals.

#### Optional arguments
##### array_template
Full path of file to use as template for the SNP array in [bed format](http://bedtools.readthedocs.io/en/latest/content/general-usage.html).
The third column is used as the physical positions of the SNP for the pseudo array. 

e.g.
```text
chr22	0	15929526
chr22	0	15991515
chr22	0	16288162
chr22	0	16926611
chr22	0	16990146
chr22	0	17498992
chr22	0	17540297
chr22	0	17728199
chr22	0	17760714
chr22	0	18180154
chr22	0	18217275
chr22	0	18220413
```


##### germline
Use [GERMLINE](https://github.com/sgusev/GERMLINE) to find shared IBD segments between all simulated individuals from pseudo array.
Does not use the genetic map to run GERMLINE.  
Runs GERMLINE as:  
```
bash ./bin/phasing_pipeline/gline.sh ./bin/germline-1-5-1/germline  ped_name map_name out_name "-bits 10 -min_m min_m"
```

##### output_dir


## Common Errors
Number of simulated segregating sites less than number of sites on template array. Increase size of simulated locus.